tests/unit/lms/views/exceptions_test.py

```diff
@@ -1,112 +1,138 @@
+class TestHTTPClientError:
+    def test_it(self, assert_response, exception, exception_views):
+        assert_response(
+            exception_views.http_client_error(),
+            status=exception.status_int,
+            message="http_client_error_test_explanation",
+        )
 
-        assert result == self.expected_result
+    @pytest.fixture
+    def exception(self):
+        return httpexceptions.HTTPBadRequest("http_client_error_test_explanation")
 
 
-class TestNotFound(ExceptionViewTest):
-    view = exceptions.notfound
-    exception = None
+class TestHAPIError:
+    def test_it(self, assert_response, exception_views):
+        assert_response(
+            exception_views.hapi_error(), 500, message="hapi_error_test_explanation"
+        )
 
-    response_status = 404
-    expected_result = {"message": "Page not found"}
+    @pytest.fixture
+    def exception(self):
+        return HAPIError("hapi_error_test_explanation")
 
 
-class TestForbidden(ExceptionViewTest):
-    view = exceptions.forbidden
-    exception = None
+class TestValidationError:
+    def test_it(self, assert_response, exception, exception_views):
+        assert_response(
+            exception_views.validation_error(),
+            status=exception.status_int,
+            error=exception,
+        )
 
-    response_status = 403
-    expected_result = {"message": "You're not authorized to view this page"}
+    @pytest.fixture
+    def exception(self):
+        return ValidationError(sentinel.messages)
 
 
-class TestHTTPClientError(ExceptionViewTest):
-    view = exceptions.http_client_error
-    exception = HTTPBadRequest("This is the error message")
+class TestReusedToolGUIDError:
+    def test_it(self, assert_response, exception_views, pyramid_request):
+        assert_response(exception_views.reused_tool_guid_error(), 200)
 
-    response_status = 400
-    expected_result = {"message": exception.args[0]}
+        pyramid_request.context.js_config.enable_error_dialog_mode.assert_called_with(
+            error_code=JSConfig.ErrorCode.REUSED_TOOL_GUID,
+            error_details={
+                "existing_tool_consumer_guid": sentinel.existing_guid,
+                "new_tool_consumer_guid": sentinel.new_guid,
+            },
+        )
 
+    @pytest.fixture
+    def exception(self):
+        return ReusedConsumerKey(sentinel.existing_guid, sentinel.new_guid)
 
-class TestHAPIError(ExceptionViewTest):
-    view = exceptions.hapi_error
-    exception = HAPIError("This is the error message")
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.context = create_autospec(
+            LTILaunchResource,
+            instance=True,
+            spec_set=True,
+            js_config=create_autospec(
+                JSConfig,
+                instance=True,
+                spec_set=True,
+                ErrorCode=JSConfig.ErrorCode,
+            ),
+        )
+        return pyramid_request
```

Comment:
The other suggested changes (inlining the `exception` and `exception_views` fixtures and grouping all the tests together into one test class) work fine with the unmodified `pyramid_request` fixture: the `pyramid_request.context` (mock `LTILaunchResource` with a mock `JSConfig`) isn't *needed* for any of the other tests, but it doesn't break any of them either.

I think it does two harms, though:

1. A bunch of unnecessary mock creation and autospeccing is getting done for every test, which slows the tests down (by a few milliseconds)
2. It's not clear that the whole mock context thing is only needed for `test_reused_tool_guid_error()` and is totally irrelevant to the rest of the tests

I think the best solution to both problems is to use two test classes:

```python
class TestExceptionViews:
    def test_notfound(self, assert_response, pyramid_request):
        ...

    def test_forbidden(self, assert_response, pyramid_request):
        ...

    ...

class TestReusedToolGUIDError:
    def test_reused_tool_guid_error(self, assert_response, pyramid_request):
        ...

    @pytest.fixture
    def pyramid_request(self, pyramid_request):
        pyramid_request.context = create_autospec(
            ...
        )
        return pyramid_request


@pytest.fixture
def assert_response(pyramid_request):
    ...
```

Now the mocking is only being done for the one test that needs it and the structure of the code also makes it clear that it only applies to that one test.

Reply:
That's what the other comment was about. You don't have to make the fixture be the pyramid request here. You can make the fixture be the context and then you don't have this coupling issue.

So you can have a fixture like this:

```python
    @pytest.fixture
    def lti_launch_resource(self, pyramid_request):
        js_config = create_autospec(
            JSConfig, instance=True, spec_set=True, ErrorCode=JSConfig.ErrorCode
        )

        return create_autospec(
            LTILaunchResource, instance=True, spec_set=True, js_config=js_config
        )
```

And then do this at the top of the single test that needs it you can have:

```python
    def test_reused_tool_guid_error(
        self, assert_response, pyramid_request, lti_launch_resource
    ):
        pyramid_request.context = lti_launch_resource
```

Then the tests can all live next to each other as happy peas in a pod!

Reply:
Sure. But I think the use of a separate test class is nice because it makes it easier to see that the `pyramid_request` / `lti_launch_resource` fixture is only relevant to one test so you don't have to worry about it when dealing with any of the other tests. Compare:

With all tests in one class you have to spot that the `lti_launch_resource` fixture is only used by one of the tests, indicated in bold:

<pre><code>class TestExceptionViews:
    def test_notfound(self, assert_response, pyramid_request):
        ...

    def test_forbidden(self, assert_response, pyramid_request):
        ...

    def test_http_client_error(self, assert_response, pyramid_request):
        ...

    def test_hapi_error(self, assert_response, pyramid_request):
        ...

    def test_validation_error(self, assert_response, pyramid_request):
        ...

    def test_reused_tool_guid_error(self, assert_response, pyramid_request, <b>lti_launch_resource</b>):
        ...

    def test_error(self, assert_response, pyramid_request):
        ...

    @pytest.fixture
    def assert_response(self, pyramid_request):
        ...

    @pytest.fixture
    def <b>lti_launch_resource</b>(self, pyramid_request):
        ...</code></pre>

Whereas with the tests in separate classes this is clearer:

<pre><code>class TestExceptionViews:
    def test_notfound(self, assert_response, pyramid_request):
        ...

    def test_forbidden(self, assert_response, pyramid_request):
        ...

    def test_http_client_error(self, assert_response, pyramid_request):
        ...

    def test_hapi_error(self, assert_response, pyramid_request):
        ...

    def test_validation_error(self, assert_response, pyramid_request):
        ...

    def test_error(self, assert_response, pyramid_request):
        ...


class TestReusedToolGUIDError:
    def test_reused_tool_guid_error(self, assert_response, <b>pyramid_request</b>):
        ...

    @pytest.fixture
    def <b>pyramid_request</b>(self, pyramid_request):
        ...

@pytest.fixture
def assert_response(pyramid_request):
    ...</code></pre>

Since the fixture is only used by one test I think there is a third alternative which is to just inline it into that test:

```python
class TestExceptionViews:

    ...

    def test_reused_tool_guid_error(self, assert_response, pyramid_request):
        pyramid_request.context = mock.create_autospec(
            ...
        )

        ...
```

Although it's kind of nice how pytest separates the creation of test fixtures from the tests themselves, so :man_shrugging: